adr/ADR-40.md

```diff
@@ -473,6 +474,24 @@ a command. This is followed by a disconnect.
 - `maximum account active connections exceeded` not notified to the client, the
   client connecting will be disconnected (seen as a connection refused.)
 
+### Client Statistics
+Clients should implement statistics gathered throughout the lifetime of the client (persisted between connections):
+
+#### Bytes In
+Should account for everything received from the server, a record of bytes received on the socket.
+
+#### Bytes Out
+Should account for everything sent to the server, a record of bytes sent on the socket.
+
+#### Messages In
+Every message (`MSG` and `HMSG`) received from the server.
+
+#### Messages out
+Every message (`PUB` and `HPUB`) sent to the server.
+
+#### Connets
```

Comment:
reconnects?

Reply:
No, I meant connects.
I feel that counting only reconnects gets messy when you have `allow initial reconnect`, which starts with 0 reconnects and 0 connets. Then, after initial connection is established, it would still be 0 reconnects, but 1 connect. Hence I feel it's clearer if we count every successful connect, including the initial one.

Reply:
We should not model the `Statistics` after Go client, or any other client, as each is counting bytes in/out in their own way. That's because the feature predates ADR and parity concepts.

We agreed that we should do what Server does - count all bytes in and out. Go counts just headears + payloads.

Regarding the connect versus reconnect - maybe we could allow this to be client specific? It can make sense depending on how `retry_on_connect` is implemented in given language after all.

Reply:
Its not new behaviour that is not align with clients. Every client counts things diferently: 
- counting just payloads as bytes in/out
- counting headers + payloads (Go does that)
- counting subjects + headers + payloads (rust)
- counting everything that goes through the socket (server & Javascript)

additionally, some languages have more fields than others.

As you can see, its impossible to make consistent ADR without introducing changes everywhere.
Because of that, we decided to count bytes as server does, which also seems to make most sense.

I agree that the `connected` vs `reconnected` might be a change we should avoid in some clients, depending how they handle `retry_initial_connect`, so I'll update the ADR that both are ok.

Reply:
@wallyqs I think that Piotr's idea is the best here:

he proposed that in clients that count data differently, we could add `total_bytes_in` `total_bytes_out` etc.
Check the upates to this PR.